query.go

```diff
@@ -103,14 +122,339 @@ func DoneChan(doneChan chan<- struct{}) QueryOption {
+		// of the same peer being asked for the same query every time.
+		lastFailed, firstUnfinished, handleQuery := -1, 0, -1
+
+		for firstUnfinished < len(queryMsgs) {
+			select {
+			case <-quit:
+				return
+			default:
+			}
+
+			handleQuery = -1
+
+			for i := firstUnfinished; i < len(queryMsgs); i++ {
+				// If we last failed at this query, skip it.
+				if i == lastFailed {
+					continue
+				}
+
+				// If this query is finished and we're at
+				// firstUnfinished, update firstUnfinished.
+				if i == firstUnfinished && atomic.
+					LoadUint32(&queryStates[i]) == 2 {
+					firstUnfinished++
+					continue
+				}
+
+				// We check to see if the query is waitng to
+				// be handled. If so, we mark it as being
+				// handled. If not, we move to the next one.
+				if !atomic.CompareAndSwapUint32(&queryStates[i],
+					0, 1) {
+					continue
+				}
+
+				// The query is now marked as in-process. We
+				// begin to process it.
+				handleQuery = i
+				sp.QueueMessageWithEncoding(queryMsgs[i],
+					nil, qo.encoding)
+			}
+
+			// Regardless of whether we have a query or not, we
+			// need a timeout.
+			timeout := time.After(qo.timeout)
+			if handleQuery == -1 {
+				if firstUnfinished == len(queryMsgs) {
+					// We've now answered all the queries.
+					return
+				}
+
+				// We have nothing to work on but not all
+				// queries are answered yet. Wait for a query
+				// timeout, or a quit signal, then see if
+				// anything needs our help.
+				select {
+				case <-quit:
+					return
+				case <-timeout:
+					if sp.Connected() {
+						continue
+					} else {
+						return
+					}
+				}
+			}
+
+			// We have a query we're working on.
+			select {
+			case <-quit:
+				return
+			case <-timeout:
+				// We failed, so set the query state back to
+				// zero and update our lastFailed state.
+				atomic.StoreUint32(&queryStates[handleQuery],
+					0)
+				if !sp.Connected() {
+					return
+				}
+				lastFailed = handleQuery
+			case m := <-matchSignal:
+				// We got a match signal. If it matches our
+				// query, we can mark this query a success.
+				if m == queryMsgs[handleQuery] {
+					atomic.StoreUint32(
+						&queryStates[handleQuery], 2)
+				}
+			}
+		}
+	}
+
+	// peerQuits holds per-peer quit channels so we can kill the goroutines
+	// when they disconnect.
+	peerQuits := make(map[*ServerPeer]chan struct{})
+
+	// matchSignals holds per-peer answer channels that get a notice that
+	// the query got a match. If it's the peer's match, the peer can
+	// mark the query a success and move on to the next query ahead of
+	// timeout.
+	matchSignals := make(map[*ServerPeer]chan wire.Message)
+
+	// Clean up on exit.
+	defer func() {
+		for _, quitChan := range peerQuits {
+			close(quitChan)
+		}
+	}()
+
+	for {
+		// Update our view of peers, starting new workers for new peers
+		// and removing disconnected/banned peers.
+		for _, peer := range s.Peers() {
```

Comment:
Alternatively, so add a bit more synchronization such that we _only_ get connected peers ever. Or perhaps, grab the mutex ourselves during this loop?